File: Mandatory/012 - Min Stack/MinStack.rb
class Node
  attr_accessor :value, :next      

  def initialize(value)
    @value = value
    @next = nil
  end
end

class LinkedList
  attr_accessor :first, :last, :count

  def initialize
    @first = nil
    @last = nil    
    @count = 0
  end

  def add(value)       
    if @last
      new_node = Node.new(value)
      @last.next = new_node
      @last = @last.next
    else
      @first = Node.new(value)
      @last = @first
    end
    @count += 1
  end

  def add_at(index, value)
    if @last
      if index == 0
        new_node = Node.new(value)
        new_node.next = @first      
        @first = new_node
        @count += 1
      elsif index == @count
        new_node = Node.new(value)
        @last.next = new_node
        @last = new_node
        @count += 1
      elsif index > 0 && index < @count
        new_node = Node.new(value)
        previous_node = get_node(index)
        new_node.next = previous_node.next
        previous_node.next = new_node
        @count += 1
      end
    elsif index == 0
      @first = Node.new(value)
      @last = @first
      @count += 1
    end    
  end

  def remove(index)
    if @last
      if index == 0
        old_node = @first
        @first = @first.next
        old_node = nil
        @count -= 1
      elsif index == @count - 1
        old_node = @last
        @last = get_node(index)
        old_node = nil
        @count -= 1
      elsif index > 0 && index < @count
        previous_node = get_node(index)
        old_node = previous_node.next
        previous_node.next = old_node.next
        old_node = nil
        @count -= 1
      end    
    end

    if @count == 0
      @first = nil
      @last = nil
    elsif @count == 1
      @first = @last if @first
      @last = @first if @last
    end
  end

  def get(index)
    node = get_node(index)
    return node.value if node
    nil
  end

  private

  def get_node(index)
    if @first
      if index >= 0 && index < @count      
        current_node = @first
        index.times { |item| current_node = current_node.next }
        current_node
      else
        nil
      end
    else
      nil
    end
  end
end

class Stack
  attr_accessor :list, :list_mins

  def initialize
    @list = LinkedList.new
    @list_mins = LinkedList.new
  end

  def push(value)
    if @list.first
      @list.add_at(0, value)
      @list_mins.add_at(0, value) if value < @list_mins.first.value
    else
      @list.add(value)
      @list_mins.add(value)
    end    
  end

  def pop
    popped = -1
    if @list.first
      popped = @list.get(0)      
      @list.remove(0)

      @list_mins.remove(0) if popped == @list_mins.first.value
    end
    popped
  end

  def min
    @list_mins.first.value
  end
end



stack = Stack.new
stack.push(3)
stack.push(5)
puts stack.min
# => 3

stack.pop
stack.push(7)
puts stack.min
# => 3

stack.push(2)
puts stack.min
# => 2

stack.pop
puts stack.min
# => 3

File: Mandatory/012 - Min Stack/MinStack.js
class Node {
  constructor(value) {
    this.value = value;
    this.next = null;
  }
}

class LinkedList {
  constructor() {
    this.first = null;
    this.last = null;
    this.count = 0;
  }

  get_node(index) {
    if (this.first) {
      if (index >= 0 && index < this.count) {
        let current_node = this.first;
        for (let i = 0; i < index; i += 1) { current_node = current_node.next }

        return current_node;
      } else {
        return null;
      }
    } else {
      return null;
    }
  }

  add(value) {
    if (this.last) {
      let new_node = new Node(value);
      this.last.next = new_node;
      this.last = this.last.next;
    } else {
      this.first = new Node(value);
      this.last = this.first
    }
    this.count += 1;
  }

  add_at(index, value) {
    let new_node;
    let previous_node;

    if (this.last) {
      if (index === 0) {
        new_node = new Node(value);
        new_node.next = this.first;
        this.first = new_node;
        this.count += 1;
      } else if (index === this.count) {
        new_node = new Node(value);
        this.last.next = new_node;
        this.last = new_node;
        this.count += 1;
      } else if (index > 0 && index < this.count) {
        new_node = new Node(value);
        previous_node = this.get_node(index);
        new_node.next = previous_node.next;
        previous_node.next = new_node;
        this.count += 1;
      }
    } else {
      this.first = new Node(value);
      this.last = this.first;
      this.count += 1;
    }
  }

  remove(index) {
    let old_node;

    if (this.last) {
      if (index === 0) {
        old_node = this.first;
        this.first = this.first.next;
        old_node = null;
        this.count -= 1;
      } else if (index === this.count - 1) {
        old_node = this.last;
        this.last = this.get_node(index);
        old_node = null;
        this.count -= 1
      } else if (index > 0 && index < this.count) {
        previous_node = this.get_node(index);
        old_node = previous_node.next;
        previous_node.next = old_node.next;
        old_node = null;
        this.count -= 1;
      }
    }

    if (this.count === 0) {
      this.first = null;
      this.last = null;
    } else if (this.count === 1) {
      if (this.first) { this.first = this.last; }
      if (this.last) { this.last = this.first; }
    }
  }

  get(index) {
    const node = this.get_node(index);
    if (node) {
      return node.value;
    } else {
      return null;
    }
  }
}

class Stack {
  constructor() {
    this.list = new LinkedList();
    this.list_mins = new LinkedList();
  }

  push(value) {
    if (this.list.first) {
      this.list.add_at(0, value);
      if (value < this.list_mins.first.value) { this.list_mins.add_at(0, value); }
    } else {
      this.list.add(value);
      this.list_mins.add(value);
    }
  }

  pop() {
    let popped = -1;
    if (this.list.first) {
      popped = this.list.get(0);
      this.list.remove(0);

      if (popped === this.list_mins.first.value) { this.list_mins.remove(0); }
    }
    return popped;
  }

  min() {
    return this.list_mins.first.value;
  }
}

const stack = new Stack();
stack.push(3);
stack.push(5);
console.log(stack.min());
// => 3

stack.pop();
stack.push(7);
console.log(stack.min());
// => 3

stack.push(2);
console.log(stack.min());
// => 2

stack.pop();
console.log(stack.min());
// => 3